sel/src/Sel/PublicKey/Internal/Signature.hs

```diff
@@ -0,0 +1,499 @@
+  Base16.extractBase16 . Base16.encodeBase16' $
+    ByteString.fromForeignPtr0
+      (Foreign.castForeignPtr publicKeyPtr)
+      (fromIntegral cryptoSignPublicKeyBytes)
+
+-- | A secret key of size 'cryptoSignSecretKeyBytes'. Keep this private.
+--
+-- @since 0.0.1.0
+newtype SecretKey = SecretKey (ForeignPtr CUChar)
+  deriving
+    ( Display
+      -- ^ @since 0.0.3.0
+      -- > display secretKey == "[REDACTED]"
+    )
+    via (OpaqueInstance "[REDACTED]" SecretKey)
+
+-- | By constant-time comparison of pointer contents.
+--
+-- @since 0.0.3.0
+instance Eq SecretKey where
+  a == b =
+    foreignPtrEqConstantTime (coerce a) (coerce b) cryptoSignSecretKeyBytes
+
+-- | > show secretKey == "[REDACTED]"
+--
+-- @since 0.0.3.0
+instance Show SecretKey where
+  show _ = "[REDACTED]"
+
+-- | Decode a hexadecimal-encoded 'StrictByteString' to a t'SecretKey'
+-- using the default copying decoder.
+--
+-- @since 0.0.3.0
+decodeSecretKeyHexByteString :: StrictByteString -> Either KeyMaterialDecodeError SecretKey
+decodeSecretKeyHexByteString bytes = SecretKey <$> copyHexKey cryptoSignSecretKeyBytes bytes
+
+-- | Encode an t'UnsafeSecretKey' to a hexadecimal encoded
+-- 'StrictByteString' using the default copying encoder.
+--
+-- @since 0.0.3.0
+encodeSecretKeyHexByteString :: UnsafeSecretKey -> StrictByteString
+encodeSecretKeyHexByteString (UnsafeSecretKey (SecretKey secretKeyPtr)) =
+  Base16.extractBase16 . Base16.encodeBase16' $
+    ByteString.fromForeignPtr0
+      (Foreign.castForeignPtr secretKeyPtr)
+      (fromIntegral cryptoSignSecretKeyBytes)
+
+-- | Produce the t'PublicKey' from a t'SecretKey'.
+--
+-- This function may throw a t'PublicKeyExtractionException' if the operation fails.
+publicKey :: SecretKey -> PublicKey
+publicKey (SecretKey secretKeyPtr) = unsafeDupablePerformIO $ do
+  publicKeyPtr <- Foreign.mallocForeignPtrBytes (fromIntegral cryptoSignPublicKeyBytes)
+  res <-
+    useM $
+      cryptoSignED25519SkToPk
+        <$> foreignPtr publicKeyPtr
+        <*> foreignPtr secretKeyPtr
+  unless (res == 0) $ throw PublicKeyExtractionException
+  pure $ PublicKey publicKeyPtr
+
+-- | Signal your intent to encode secret key material for transmission
+-- by wrapping a t'SecretKey' in t'UnsafeSecretKey'.
+--
+-- @since 0.0.3.0
+newtype UnsafeSecretKey = UnsafeSecretKey SecretKey
+  deriving newtype
+    ( Eq
+      -- ^ @since 0.0.3.0
+      -- Follows the t'SecretKey' instance.
+    )
+  deriving
+    ( Display
+      -- ^ @since 0.0.3.0
+      -- Hexadecimal-encoded bytes.
+    )
+    via (ShowInstance UnsafeSecretKey)
+
+-- | Hexadecimal-encoded bytes.
+--
+-- @since 0.0.3.0
+instance Show UnsafeSecretKey where
+  show = ByteString.unpackChars . encodeSecretKeyHexByteString
+
+-- | By lexicographical comparison of pointer contents.
+--
+-- ⚠️ Vulnerable to timing attacks!
+--
+-- @since 0.0.3.0
+instance Ord UnsafeSecretKey where
+  a `compare` b =
+    foreignPtrOrd (coerce a) (coerce b) cryptoSignSecretKeyBytes
```

Comment:
This feels a bit over-engineered. `encodeSecretKeyHexByteString` ought to be a much simpler interface, no? I don't understand if there are invariants that are preserved by this type and its instances.

Reply:
It's intended as a guardrail to prevent you from doing things like serializing secret keys and using `Ord` in a way that's vulnerable to timing attacks without some kind of signal to readers that you're doing something that demands some extra attention.

Reply:
We do already have a (albeit low-tech) mechanism for signalling unsafe operations: 
![image](https://github.com/user-attachments/assets/af5386b8-1add-4e41-99bc-36804e7b8e5a)

Reply:
I tend to prefer sticking such things behind a gate that's very slightly higher than that since `unsafe` is a very overloaded word. To my mind, it implies there's some `safeSecretKeyToHexByteString` waiting to be discovered, and there just isn't. Disjoint concepts <=> disjoint types. I'm not married to it, though, and with constant-time `Ord` coming quite soon I could be convinced the power-to-weight isn't there.

Reply:
That option _is_ available in the present API as `unsafeSecretKeyHexByteString`, which just wraps whatever `SecretKey` in the newtype first.

Reply:
Pushed commits to strip that all out. Will update the API golden test once #171 lands and I rebase this PR on top of it.